### PR TITLE
Find first and last detectorID's in a component

### PR DESCRIPTION
**Pull request by jmborr**
**Description of work.**  

Companion to PR #32657
For details, see [REF81](https://code.ornl.gov/sns-hfir-scse/diffraction/engineering-diffraction/engineering-diffraction/-/issues/81)

A component may have its detectorID's unsorted, thus we cannot rely on the first and last listed detectors as those with the smallest and largest detector-ID values.

- [x] correct function to find first and last detector-ID's
- [x] integration test
- [x] release notes

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/AlignComponents.py [hunk 1]
```diff
@@ -623,30 +627,37 @@ def _minimisation_func(self, x_0, wks_name, component, firstIndex, lastIndex):
...
-
-        @param component :: reference to a detector component object
-
-        @returns detector ID (`int`) of the last detector in the component        """
-        if component.type() == 'DetectorComponent' or component.type() == 'GridDetectorPixel':
-            return component.getID()
-        else:
-            return self._getLastDetID(component[component.nelements() - 1])
+        unique_name = self._unique_name(component, component_info)
+        index = component_info.indexOfAny(unique_name)  # component-info index
```
Comment (peterfpeterson) [inline Framework/PythonInterface/plugins/algorithms/AlignComponents.py:654]: What happens if the name doesn't exist in the instrument?

File: Testing/SystemTests/tests/framework/AlignComponentsTest.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                            PeakPositions='1.3143, 1.3854,1.6967, 1.8587, 2.0781',
+                            AdjustmentsTable=prefix + 'adjustments',
+                            DisplacementsTable=prefix + 'displacements',
+                            InputWorkspace=prefix + 'snap',
+                            OutputWorkspace=prefix + 'snap_modified',
+                            FitSourcePosition=False, FitSamplePosition=False,
+                            ComponentList='Column1',
+                            XPosition=False, YPosition=False, ZPosition=True,
+                            AlphaRotation=False, BetaRotation=False, GammaRotation=False)
+            assert 'First and last detectorID for Column1 are 0, 196607' in open(log_file, 'r').read()
```
Comment (peterfpeterson) [inline Testing/SystemTests/tests/framework/AlignComponentsTest.py:44]: nice!

File: Framework/PythonInterface/plugins/algorithms/AlignComponents.py (hunk 1 above)
Comment (jmborr) [reply Framework/PythonInterface/plugins/algorithms/AlignComponents.py:654]: I see what you mean. `AlignComponents` didn't check for incorrect component names but I guess we can now `raise` some screaming to that effect here.

**Review by peterfpeterson — APPROVED** [review]

File: Framework/PythonInterface/mantid/utils/logging.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+# standard packages
+from contextlib import contextmanager
+import os
+import sys
+import tempfile
+
+
+@contextmanager
+def to_file(level=None, filename=None):
```
Comment (AndreiSavici) [inline Framework/PythonInterface/mantid/utils/logging.py:19]: How is this different than setting the file channel?

File: Framework/PythonInterface/plugins/algorithms/AlignComponents.py [hunk 4]
```diff
@@ -623,30 +627,37 @@ def _minimisation_func(self, x_0, wks_name, component, firstIndex, lastIndex):
         # calculate the fractional peak center deviations, then sum their absolute values
         return np.sum(np.abs((peaks_d - self.peaks_ref) / self.peaks_ref))
 
-    def _getFirstDetID(self, component):
-        """
-        recursive search to find first detID of a component
+    def _unique_name(self, component, component_info):
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/AlignComponents.py:630]: Can you make this functions belong to the instrument? I would like to use them somewhere else

File: Framework/PythonInterface/mantid/utils/logging.py (hunk 3 above)
Comment (jmborr) [reply Framework/PythonInterface/mantid/utils/logging.py:19]: what file channel are you referring to?

File: Framework/PythonInterface/plugins/algorithms/AlignComponents.py (hunk 4 above)
Comment (jmborr) [reply Framework/PythonInterface/plugins/algorithms/AlignComponents.py:630]: Please make a Story for this functionality
```c++
vector<size_t> Instrument::detectorIDs(std::string componentName) const;
```

File: Framework/PythonInterface/mantid/utils/logging.py (hunk 3 above)
Comment (jmborr) [reply Framework/PythonInterface/mantid/utils/logging.py:19]: passing `FileChannel` to the configuration service is not understood, meaning the the `Poco` file channel is not set, hence the trick to redirect `sys.stdout` to write to a file

**Pull request merged**